addons/html_builder/static/src/website_sale/products_item_option_plugin.js

```diff
@@ -0,0 +1,430 @@
+      ribbonElement.style.color = ribbon.text_color || "";
+    });
+
+    return save ? this._saveRibbons() : "";
+  }
+  /**
+   * Returns all ribbon classes, current and deleted, so they can be removed.
+   *
+   */
+  _getRibbonClasses() {
+    return (
+      Object.values(this.ribbons).reduce((classes, ribbon) => {
+        return classes + ` ${this.ribbonPositionClasses[ribbon.position]}`;
+      }, "") + this.deletedRibbonClasses
+    );
+  }
+
+  async _saveRibbons() {
+    const originalIds = Object.keys(this.originalRibbons).map((id) =>
+      parseInt(id)
+    );
+    const currentIds = this.ribbons.map((ribbon) => parseInt(ribbon.id));
+
+    const created = this.ribbons.filter(
+      (ribbon) => !originalIds.includes(ribbon.id)
+    );
+    const deletedIds = originalIds.filter((id) => !currentIds.includes(id));
+    const modified = this.ribbons.filter((ribbon) => {
+      if (created.includes(ribbon)) {
+        return false;
+      }
+      const original = this.originalRibbons[ribbon.id];
+      return Object.entries(ribbon).some(
+        ([key, value]) => value !== original[key]
+      );
+    });
+
+    const proms = [];
+    let createdRibbonIds;
+    if (created.length > 0) {
+      proms.push(
+        this.services.orm
+          .create(
+            "product.ribbon",
+            created.map((ribbon) => {
+              ribbon = Object.assign({}, ribbon);
+              this.originalRibbons[ribbon.id] = ribbon;
+              delete ribbon.id;
+              return ribbon;
+            })
+          )
+          .then((ids) => (createdRibbonIds = ids))
+      );
+    }
+
+    modified.forEach((ribbon) => {
+      const ribbonData = {
+        name: ribbon.name,
+        bg_color: ribbon.bg_color,
+        text_color: ribbon.text_color,
+        position: ribbon.position,
+      };
+      proms.push(
+        this.services.orm.write("product.ribbon", [ribbon.id], ribbonData)
+      );
+      this.originalRibbons[ribbon.id] = Object.assign({}, ribbon);
+    });
+
+    if (deletedIds.length > 0) {
+      proms.push(this.services.orm.unlink("product.ribbon", deletedIds));
+    }
+
+    await Promise.all(proms);
+
+    const localToServer = Object.assign(
+      this.ribbonsObject,
+      Object.fromEntries(
+        created.map((ribbon, index) => [
+          ribbon.id,
+          { ...this.ribbonsObject[ribbon.id], id: createdRibbonIds[index] },
+        ])
+      ),
+      {
+        false: {
+          id: "",
+        },
+      }
+    );
+
+    // Building the final template to ribbon-id map
+    const finalTemplateRibbons = this.productTemplatesRibbons.reduce(
+      (acc, { templateId, ribbonId }) => {
+        acc[templateId] = ribbonId;
+        return acc;
+      },
+      {}
+    );
+    // Inverting the relationship so that we have all templates that have the same ribbon to reduce RPCs
+    const ribbonTemplates = Object.entries(finalTemplateRibbons).reduce(
```

Comment:
i know that some people like a functional style, but to be honest, i think that in many cases, a simple for loop is easier to maintain and to debug. it also has about the same number of lines of code
```js
const ribbonTemplates = {};
for (const templateId in finalTemplateRibbons) {
  const ribbonId = finalTemplateRibbons[templateId];
  if (!ribbonTemplates[ribbonId]) {
    ribbonTemplates[ribbonId] = [];
  }
  ribbonTemplates[ribbonId].push(parseInt(templateId));
}
```